typemap: attempt split on typenames also

vtjnash opened the pull request:
This is a recreation of https://github.com/JuliaLang/julia/pull/16418, now working due to many changes in the meantime! In measurements, it appears to successfully shave significant time off `require` calls:
```
$ time ./usr/bin/julia -e 'using DataFrames'
user 0m1.192s => 0m0.887s

$ time ./usr/bin/julia -e 'using Plots'
user 0m3.240s => 0m2.056s

$ time ./usr/bin/julia -e 'using ApproxFun
user 0m21.551s => 0m11.953s
```

vtjnash committed:
make the TypeMap capable of splitting using the TypeName group of the argument

Previously, the TypeMap could only split on leaf types. Now it also has
the ability to split on more abstract types.

This helps gives types their own linear list to scan to find its
constructors, instead of needing to search through all constructors.

vtjnash committed:
typemap: small correction against forming an invalid query

If the parameter is a TypeVar, it cnan have subtypes that might appear
in the map itself and require a scan to locate.

JeffBezanson commented on src/typemap.c:
```diff
@@ -851,23 +1140,60 @@ static void jl_typemap_level_insert_(
+    else {
+        t1 = NULL;
+    }
+    if (0 && t1 == NULL) {
```
?

JeffBezanson commented on src/typemap.c:
```diff
@@ -19,9 +19,80 @@ extern "C" {
+        jl_datatype_t *dt = (jl_datatype_t*)t1;
+        if ((abstract || !dt->abstract) && !jl_is_kind(t1))
+            return 0;
+        return 1;
```
In this case we return `NULL` above, so in what sense is that imprecise? Or maybe a better question is, when does this function differ from `jl_type_extract_name() == NULL`?

vtjnash replied:
I suppose NULL is imprecise in the sense that we'll never consider asking this `jl_type_extract_name_imprecise` question if the result was NULL?

JeffBezanson commented on src/typemap.c:
```diff
@@ -280,16 +375,24 @@ static int jl_typemap_intersection_array_visitor(jl_array_t *a, jl_value_t *ty,
-            (tparam
-             ? (jl_typeof(t) == ty || jl_isa(t, ty)) // (Type{t} <: ty), where is_leaf_type(t) => isa(t, ty)
-             : (t == ty || jl_subtype(t, ty)))) {
+        if (tparam & 2) {
```
Could use a comment on this function saying what the bits of `tparam` mean.

vtjnash replied:
Whether it's the TypeName and whether it's the parameter of Type{} (but agree it should have a comment)

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

vtjnash commented:
```
$ time ./julia -e '@time using Plots; @time display(@time plot(1:10))'
  4.815404 seconds (11.86 M allocations: 742.897 MiB, 4.46% gc time)
  1.499383 seconds (2.45 M allocations: 139.535 MiB, 2.96% gc time)
  4.883966 seconds (8.14 M allocations: 457.537 MiB, 2.23% gc time)

real	0m10.145s
user	0m9.966s
sys	0m0.384s
```

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/35384#issuecomment-655602685) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/8919eb56fb36ab9b1a6e403d7940607f460dfa37/2931ff9_vs_576769a/report.md). cc @ararslan

vtjnash committed:
typemap: handle typeof(Union{}) as equivalent to Type{Union{}}

This is an odd inversion in the specificity lattice where supertype(T)
is more specific than T (and type-equal to T). Handle that here by
consistently normalizing to the more specific copy, to avoid
complications later.
